### PR TITLE
crimson: retain map references in OSDSingletonState::store_maps

### PR DESCRIPTION
**Pull request by athanatos**
Includes some logging improvements helpful to debugging.

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by Matan-B — COMMENTED** [review]
LGTM, I had a similar branch (LocalOSDMapRef usage) but haven't yet encountered the crashes mentioned in the tracker.

Note (out of this PR scope):
* We should probably let store_maps return added_maps to be used later by (unimplemented) `track_pools_and_pg_num_changes`. This method will be responsible for both tracking pg num changes and `store_final_pool_info`.

CC: @amathuria

File: src/crimson/osd/shard_services.cc [hunk 1]
```diff
@@ -500,19 +500,27 @@ seastar::future<> OSDSingletonState::store_maps(ceph::os::Transaction& t,
...
-      meta_coll->store_final_pool_info(t, lastmap, added_maps);
-      return seastar::now();
+      epoch_t last_map_epoch = superblock.get_newest_map();
+      auto last_map_fut = last_map_epoch > 0
+	? get_local_map(last_map_epoch)
+	: seastar::make_ready_future<local_cached_map_t>();
+      return last_map_fut.then(
+	[&t, FNAME, last_map_epoch, this, &added_maps](auto lastmap) {
+	INFO("storing final pool info lastmap epoch {}, added maps {}->{}",
+	     last_map_epoch > 0 ? lastmap->get_epoch() : 0,
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/osd/shard_services.cc:518]: By using last_map_epoch, yes.  lastmap may be null.

File: src/crimson/osd/osd_meta.cc [hunk 2]
```diff
@@ -105,31 +105,31 @@ OSDMeta::load_final_pool_info(int64_t pool) {
...
-  OSDMap* lastmap,
-  std::map<epoch_t, OSDMap*> &added_map)
+  LocalOSDMapRef previous,
+  std::map<epoch_t, LocalOSDMapRef> &added_map)
 {
   for (auto [e, map] : added_map) {
-    if (!lastmap) {
-      lastmap = map;
+    if (!previous) {
+      previous = map;
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/osd/osd_meta.cc:113]: No, previous may be null if last_map_epoch in store_maps is 0.

**Comment by athanatos** [comment]
Retesting

**Comment by athanatos** [comment]
Hmm, rebase seems to have introduced a crash -- presumably one of the logging changes.  Will debug in the morning.

**Comment by athanatos** [comment]
Crash actually due to https://tracker.ceph.com/issues/64140 introduced by https://github.com/ceph/ceph/pull/55262, revert PR https://github.com/ceph/ceph/pull/55288

Retesting.

**Review by xxhdx1985126 — APPROVED** [review]

**Review by Matan-B — APPROVED** [review]

**Comment by athanatos** [comment]
Tests passed: https://pulpito.ceph.com/sjust-2024-01-24_15:38:26-crimson-rados-wip-sjust-crimson-testing-2024-01-23-distro-default-smithi/

**Comment by athanatos** [comment]
Repushed with trivial fix for make check error (unecessary FNAME capture).

**Comment by athanatos** [comment]
jenkins test make check

**Comment by athanatos** [comment]
jenkins test api

**Pull request merged**